### PR TITLE
Optimize startup time

### PR DESCRIPTION
**Pull request by TheItivitist**
**Optimize listJar() method**

listJar() is a method that returns a list of files contained within a Jar file according to a glob pattern.

It used to rely on a ZipInputStream that needs to read the entire file to discover its content.

Now switched to a ZipFileSystem that is able to use the Central Directory at the end of the file directly to list contained files, therefore it is by order of magnitude faster.

**Optimize pathing jar generation**

Manually generate a file name using regular Random rather than using Files.createTempFile() since this latter uses SecuredRandom which itself queries network interfaces which takes more time than a good old System.nanotime() of Random.

### CODE REVIEW

**Review by benoitdesire — APPROVED** [review]

**Review by benoitdesire — APPROVED** [review]

**Review by xdaragon — APPROVED** [review]

**Review by xdaragon — APPROVED** [review]

**Review by xdaragon — APPROVED** [review]

File: capsule-util/src/test/java/co/paralleluniverse/capsule/CapsuleLauncherTest.java [hunk 1]
```diff
@@ -50,7 +52,7 @@ public void testSimpleExtract() throws Exception {
                 .addEntry("lib/a.jar", emptyInputStream())
                 .addEntry("lib/b.class", emptyInputStream())
                 .addEntry("q/w/x.txt", emptyInputStream())
-                .addEntry("d\\f\\y.txt", emptyInputStream()) // test with Windows path
+                .addEntry("d/f/y.txt", emptyInputStream())
```
Comment (dfev77) [inline capsule-util/src/test/java/co/paralleluniverse/capsule/CapsuleLauncherTest.java:55]: windows support not needed anymore? (minor?)

File: capsule/src/main/java/Capsule.java [hunk 2]
```diff
@@ -3693,20 +3702,39 @@ private static boolean shouldExtractFile(String fileName) {
...
+                        return Files.walk(rootDir);
+                    }
+                    catch (IOException e)
+                    {
+                        throw new RuntimeException(e);
+                    }
+                })
+                .filter(p -> !regular || Files.isRegularFile(p))
+                .map(Path::toString)
+                .map(p -> p.startsWith("/") ? p.substring(1) : p)
```
Comment (dfev77) [inline capsule/src/main/java/Capsule.java:3726]: Could this be a backslash on windows?
PS: what is the purpose of this check and strip?

File: capsule/src/main/java/Capsule.java [hunk 3]
```diff
@@ -5459,7 +5460,8 @@ static Path createPathingJar(Path dir, List<Path> cp) {
         try {
             final Path absolutePath = dir.toAbsolutePath();
             final List<String> paths = createPathingClassPath(absolutePath, cp);
-            final Path pathingJar = Files.createTempFile(absolutePath, "capsule_pathing_jar", ".jar");
+
+            final Path pathingJar = createTempFile(absolutePath, "capsule_pathing_jar", ".jar");
```
Comment (dfev77) [inline capsule/src/main/java/Capsule.java:5465]: I suppose the Files.createTempFile also check for file existence and generates a new file name if exists.
The new method does not have this fallback (and I think it should)

File: capsule-util/src/test/java/co/paralleluniverse/capsule/CapsuleLauncherTest.java (hunk 1 above)
Comment (TheItivitist) [reply capsule-util/src/test/java/co/paralleluniverse/capsule/CapsuleLauncherTest.java:55]: Jar file separator (for its content) is always '/', it is not platform dependant

File: capsule/src/main/java/Capsule.java (hunk 2 above)
Comment (TheItivitist) [reply capsule/src/main/java/Capsule.java:3726]: We are browsing a Jar content, so all filenames are separated with '/' whatever platform we are working with.

Beside filenames in jars can start with a '/' or not; however the path() function turning Strings into (relative) Paths uses the default filesystem provider to build the path, so starting with a '/' can have a special meaning (absolute path starting from unix root) which we definitely don't want here.

So we normalize all  jar entries paths without the '/', to end up with the same result as the previous implementation

File: capsule/src/main/java/Capsule.java (hunk 3 above)
Comment (TheItivitist) [reply capsule/src/main/java/Capsule.java:5465]: arg indeed
will fix that

**Pull request merged**